zerver/lib/test_classes.py

```diff
@@ -279,6 +279,18 @@ def run(self, result: TestResult | None = None) -> TestResult | None:  # nocover
         token=r"[a-z0-9_]{24}"
     )
 
+    @override
+    def assertEqual(self, first: Any, second: Any, msg: Any = "") -> None:
+        if isinstance(first, str) and isinstance(second, str):
+            if first != second:
+                raise AssertionError(
+                    "Actual and expected outputs do not match; showing diff.\n"
+                    + diff_strings(first, second)
+                    + str(msg)
+                )
+        else:
+            super().assertEqual(first, second, msg)
```

Comment:
We should not need this at all. Python’s built-in `assertEqual` has been showing string diffs by default [since Python 2.7](https://docs.python.org/3/whatsnew/2.7.html#updated-module-unittest) (for Unicode strings).

Reply:
Looking more closely, this uses `web/tests/lib/mdiff.js` to produce colored output -- which is a little nicer, but I'm not sure is worth the complexity.  It dates back to 44e6d9b54bb88dab8bbbb03ef8f0ceccf349e438 in 2017.

Reply:
I seem to recall that this did solve an important practical problem for Markdown tests. From https://github.com/zulip/zulip/pull/7782:
```
An example of what the new diffs let you do that the previous do not; The longest fixture in which the javascript frontend and python backend disagree is kind of impenetrable with normal diffs, the lines are too long to navigate without color and there's a lot of irrelevant differences such as changes in attribute order. But, now it's not very difficult to see that the only real difference is that the marked output simply lacks <p> tags around <span> (which happens to be the case a lot of the time; would be really difficult to tell this is the only problem without these changes).
https://imgur.com/a/Mjvom
```

Reply:
Changed this back to Alex's previous version that preserves the colorization change. Since the adjusted code only runs when a test is failing, it seems very cheap to keep around and I have some very bad memories of hand-diffing HTML without this.